pedsnetdcc/utils.py

```diff
@@ -67,3 +76,83 @@ def make_conn_str(url, search_path=None, password=None):
             parts.append('{0}={1}'.format(k, value))
 
     return ' '.join(parts)
+
+
+def get_conn_info_dict(conn_str):
+    """Return the connection info form a libpq-compliant conn string as a dict.
+
+    The `user`, `host`, `port`, `dbname`, and `search_path` are extracted using
+    regular expressions. If any of them are not found, the corresponding dict
+    value will be None.
+
+    See https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING
+
+    :param str conn_str: a libpq-compliant connection string
+    :returns:            the extracted connection info in dict form
+    :rtype:              dict
+    """  # noqa
+
+    result = {'user': None, 'host': None, 'port': None, 'dbname': None,
+              'search_path': None}
+
+    host_match = re.search(r"host=(\S*)", conn_str)
+    port_match = re.search(r"port=(\S*)", conn_str)
+    dbname_match = re.search(r"dbname=(\S*)", conn_str)
+    user_match = re.search(r"user=(\S*)", conn_str)
+    search_path_match = re.search(r"search_path=(.*?)[' ]", conn_str)
+
+    if host_match:
+        result['host'] = host_match.group(1)
+    if port_match:
+        result['port'] = port_match.group(1)
+    if dbname_match:
+        result['dbname'] = dbname_match.group(1)
+    if user_match:
+        result['user'] = user_match.group(1)
+    if search_path_match:
+        result['search_path'] = search_path_match.group(1)
+
+    return result
+
+
+def combine_dicts(*args):
+    """Return a new dict that combines all the args in successive update calls.
+
+    :param args: any number of dict-type objects
+    :returns:    a dict which is the result of combining all the args
+    :rtype:      dict
+    """
+
+    result = {}
+
+    for arg in args:
+        result.update(arg)
+
+    return result
+
+
+def check_stmt_data(stmt, caller_name=''):
```

Comment:
I definitely had that thought, too. I ended up keeping them separate because I really want the `Statement` class and `db.py` generally to stay out of the business of deciding what to do with database errors. Also, I want to keep the API super small to facilitate sub-classing and/or straight object substitution.